File: app/src/main/java/me/eycia/picScalerView/ProgressBar.java
package me.eycia.picScalerView;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.ColorFilter;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.util.Log;

import com.facebook.drawee.drawable.DrawableUtils;

/**
 * A progress bar show in center
 */
public class ProgressBar extends Drawable {
    private final Paint mPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private int mBackgroundColor = 0x80000000;
    private int mTextColor = Color.BLACK;
    private int mColor = 0x800080FF;
    private int mPadding = 10;
    private int mBarWidth = 20;
    private int mLevel = 0;
    private boolean mHideWhenZero = false;

    public ProgressBar(int BackgroundColor, int Color, int TextColor) {
        super();
        this.mBackgroundColor = BackgroundColor;
        this.mColor = Color;
        this.mTextColor = TextColor;
    }

    /**
     * Sets the progress bar color.
     */
    public void setTextColor(int color) {
        if (mTextColor != color) {
            mTextColor = color;
            invalidateSelf();
        }
    }

    /**
     * Gets the progress bar color.
     */
    public int getTextColor() {
        return mTextColor;
    }


    /**
     * Sets the progress bar color.
     */
    public void setColor(int color) {
        if (mColor != color) {
            mColor = color;
            invalidateSelf();
        }
    }

    /**
     * Gets the progress bar color.
     */
    public int getColor() {
        return mColor;
    }

    /**
     * Sets the progress bar background color.
     */
    public void setBackgroundColor(int backgroundColor) {
        if (mBackgroundColor != backgroundColor) {
            mBackgroundColor = backgroundColor;
            invalidateSelf();
        }
    }

    /**
     * Gets the progress bar background color.
     */
    public int getBackgroundColor() {
        return mBackgroundColor;
    }

    /**
     * Sets the progress bar padding.
     */
    public void setPadding(int padding) {
        if (mPadding != padding) {
            mPadding = padding;
            invalidateSelf();
        }
    }

    /**
     * Gets the progress bar padding.
     */
    @Override
    public boolean getPadding(Rect padding) {
        padding.set(mPadding, mPadding, mPadding, mPadding);
        return mPadding != 0;
    }

    /**
     * Sets the progress bar width.
     */
    public void setBarWidth(int barWidth) {
        if (mBarWidth != barWidth) {
            mBarWidth = barWidth;
            invalidateSelf();
        }
    }

    /**
     * Gets the progress bar width.
     */
    public int getBarWidth() {
        return mBarWidth;
    }

    /**
     * Sets whether the progress bar should be hidden when the progress is 0.
     */
    public void setHideWhenZero(boolean hideWhenZero) {
        mHideWhenZero = hideWhenZero;
    }

    /**
     * Gets whether the progress bar should be hidden when the progress is 0.
     */
    public boolean getHideWhenZero() {
        return mHideWhenZero;
    }

    @Override
    protected boolean onLevelChange(int level) {
        mLevel = level;
        invalidateSelf();
        return true;
    }

    @Override
    public void setAlpha(int alpha) {
        mPaint.setAlpha(alpha);
    }

    @Override
    public void setColorFilter(ColorFilter cf) {
        mPaint.setColorFilter(cf);
    }

    @Override
    public int getOpacity() {
        return DrawableUtils.getOpacityFromColor(mPaint.getColor());
    }

    @Override
    public void draw(Canvas canvas) {
        if (mHideWhenZero && mLevel == 0) {
            return;
        }
        drawBar(canvas, 10000, mBackgroundColor);
        drawBar(canvas, mLevel, mColor);
        drawText(canvas, mLevel, mTextColor);
    }

    private void drawText(Canvas canvas, int level, int color) {
        Rect bounds = getBounds();
        String texts = String.format("%5.1f/100.0", (level + 0.0) / 100.0);

        Log.d("msghub", texts);

        mPaint.setTextAlign(Paint.Align.CENTER);
        mPaint.setTextSize(30);

        int xpos_text = bounds.width() / 2;
        int ypos_text = bounds.height() / 2;
        mPaint.setColor(color);
        canvas.drawText(texts, xpos_text, ypos_text, mPaint);
    }

    private void drawBar(Canvas canvas, int level, int color) {
        Rect bounds = getBounds();
        int length = (bounds.width() - 2 * mPadding) * level / 10000;
        int xpos = bounds.left + mPadding;
        int ypos = (bounds.height() - mBarWidth) / 2;
        mPaint.setColor(color);
        canvas.drawRect(xpos, ypos, xpos + length, ypos + mBarWidth, mPaint);
    }
}

File: app/src/main/java/me/eycia/picScalerView/ProgressBar.kt
package me.eycia.picScalerView

import android.graphics.*
import android.graphics.drawable.Drawable
import android.util.Log
import com.facebook.drawee.drawable.DrawableUtils

/**
 * A progress bar show in center
 */
class ProgressBar(BackgroundColor: Int, color: Int, TextColor: Int) : Drawable() {
    private val mPaint = Paint(Paint.ANTI_ALIAS_FLAG)

    var backgroundColor = 0x80000000.toInt()
        set(backgroundColor) {
            if (field != backgroundColor) {
                field = backgroundColor
                invalidateSelf()
            }
        }

    var textColor = Color.BLACK
        set(color) {
            if (textColor != color) {
                field = color
                invalidateSelf()
            }
        }

    var color = 0x800080FF.toInt()
        set(color) {
            if (this.color != color) {
                this.color = color
                invalidateSelf()
            }
        }

    private var mPadding = 10
    fun setPadding(padding: Int) {
        if (mPadding != padding) {
            mPadding = padding
            invalidateSelf()
        }
    }

    var barWidth = 20
        set(barWidth) {
            if (this.barWidth != barWidth) {
                this.barWidth = barWidth
                invalidateSelf()
            }
        }
    private var mLevel = 0

    var hideWhenZero = false

    init {
        this.backgroundColor = BackgroundColor
        this.color = color
        this.textColor = TextColor
    }

    /**
     * Gets the progress bar padding.
     */
    override fun getPadding(padding: Rect): Boolean {
        padding.set(mPadding, mPadding, mPadding, mPadding)
        return mPadding != 0
    }

    override fun onLevelChange(level: Int): Boolean {
        mLevel = level
        invalidateSelf()
        return true
    }

    override fun setAlpha(alpha: Int) {
        mPaint.alpha = alpha
    }

    override fun setColorFilter(cf: ColorFilter?) {
        mPaint.colorFilter = cf
    }

    override fun getOpacity(): Int {
        return DrawableUtils.getOpacityFromColor(mPaint.color)
    }

    override fun draw(canvas: Canvas) {
        if (hideWhenZero && mLevel == 0) {
            return
        }
        drawBar(canvas, 10000, backgroundColor)
        drawBar(canvas, mLevel, color)
        drawText(canvas, mLevel, textColor)
    }

    private fun drawText(canvas: Canvas, level: Int, color: Int) {
        val bounds = bounds
        val texts = String.format("%5.1f/100.0", (level + 0.0) / 100.0)

        Log.d("msghub", texts)

        mPaint.textAlign = Paint.Align.CENTER
        mPaint.textSize = 30f

        val xpos_text = bounds.width() / 2
        val ypos_text = bounds.height() / 2
        mPaint.color = color
        canvas.drawText(texts, xpos_text.toFloat(), ypos_text.toFloat(), mPaint)
    }

    private fun drawBar(canvas: Canvas, level: Int, color: Int) {
        val bounds = bounds
        val length = (bounds.width() - 2 * mPadding) * level / 10000
        val xpos = bounds.left + mPadding
        val ypos = (bounds.height() - barWidth) / 2
        mPaint.color = color
        canvas.drawRect(xpos.toFloat(), ypos.toFloat(), (xpos + length).toFloat(), (ypos + barWidth).toFloat(), mPaint)
    }
}
